h2/test_h2_headers.py

```diff
@@ -106,3 +107,192 @@ def test_random_header(self):
         self.start_tempesta()
         self.start_all_clients()
         self.wait_while_busy(curl)
+
+class CurlTestBase(tester.TempestaTest):
+
+    clients = [
+        {
+            'id' : 'curl',
+            'type' : 'external',
+            'binary' : 'curl',
+            'cmd_args' : (
+                '-kf ' # Set non-null return code on 4xx-5xx responses.
+                'https://${tempesta_ip}/ '
+                )
+        },
+    ]
+
+    def test(self):
+        curl = self.get_client('curl')
+
+        self.start_all_servers()
+        self.start_tempesta()
+
+        self.start_all_clients()
+        self.wait_while_busy(curl)
+        self.assertEqual(0, curl.returncode,
+                         msg=("Curl return code is not 0 (%d)." %
+                              (curl.returncode)))
+        curl.stop()
+
+        self.start_all_clients()
+        self.wait_while_busy(curl)
+        self.assertEqual(0, curl.returncode,
+                         msg=("Curl return code is not 0 (%d)." %
+                              (curl.returncode)))
+
+class AddBackendShortHeaders(CurlTestBase):
+    ''' The test checks the correctness of forwarding short headers with
+    duplication
+    '''
+
+    backends = [
+        {
+            'id' : 'nginx',
+            'type' : 'nginx',
+            'port' : '8000',
+            'status_uri' : 'http://${server_ip}:8000/nginx_status',
+            'config' : NGINX_CONFIG % """
+add_header x-extra-data1 "q";
+add_header x-extra-data2 "q";
+add_header x-extra-data1 "q";
+
+return 200;
+""",
+        }
+    ]
+
+    tempesta = {
+        'config' : TEMPESTA_CONFIG % "",
+    }
+
+    def test(self):
+        CurlTestBase.test(self)
+        nginx = self.get_server('nginx')
+        nginx.get_stats()
+        self.assertEqual(2, nginx.requests)
+
+class AddBackendShortHeadersCache(CurlTestBase):
+    ''' The test checks the correctness of serving short headers with duplicate
+    from the cache
+    '''
+
+    backends = [
+        {
+            'id' : 'nginx',
+            'type' : 'nginx',
+            'port' : '8000',
+            'status_uri' : 'http://${server_ip}:8000/nginx_status',
+            'config' : NGINX_CONFIG % """
+add_header x-extra-data1 "q";
+add_header x-extra-data2 "q";
+add_header x-extra-data1 "q";
+
+return 200;
+""",
+        }
+    ]
+
+    tempesta = {
+        'config' : TEMPESTA_CONFIG % "cache_fulfill * *;",
+    }
+
+    def test(self):
+        CurlTestBase.test(self)
+        nginx = self.get_server('nginx')
+        nginx.get_stats()
+        self.assertEqual(1, nginx.requests)
```

Comment:
First I wanted to say that it's better to check counters on Tempesta side. But I see your intention, backend counters won't lie about serving from cache, while system under test possibly can. So Ok, lets leave it as is.